src/theory/bags/bags_rewriter.cpp

```diff
@@ -0,0 +1,427 @@
+/*********************                                                        */
+/*! \file bags_rewriter.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Mudathir Mohamed
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Bags theory rewriter.
+ **/
+
+#include "theory/bags/bags_rewriter.h"
+
+#include "normal_form.h"
+
+using namespace CVC4::kind;
+
+namespace CVC4 {
+namespace theory {
+namespace bags {
+
+RewriteResponse BagsRewriter::postRewrite(TNode n)
+{
+  if (n.isConst())
+  {
+    // no need to rewrite n if it is already in a normal form
+    return RewriteResponse(REWRITE_DONE, n);
+  }
+  if (NormalForm::AreChildrenConstants(n))
+  {
+    Node normal = NormalForm::getNormalForm(n);
+    return RewriteResponse(REWRITE_DONE, n);
+  }
+  Kind k = n.getKind();
+  switch (k)
+  {
+    case MK_BAG: return rewriteMakeBag(n);
+    case BAG_COUNT: return rewriteBagCount(n);
+    case UNION_MAX: return rewriteUnionMax(n);
+    case UNION_DISJOINT: return rewriteUnionDisjoint(n);
+    case INTERSECTION_MIN: return rewriteIntersectionMin(n);
+    case DIFFERENCE_SUBTRACT: return rewriteDifferenceSubtract(n);
+    case DIFFERENCE_REMOVE: return rewriteDifferenceRemove(n);
+    case BAG_CHOOSE: return rewriteChoose(n);
+    case BAG_CARD: return rewriteCard(n);
+    case BAG_IS_SINGLETON: return rewriteIsSingleton(n);
+    default: break;
+  }
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::preRewrite(TNode n)
+{
+  Kind k = n.getKind();
+  switch (k)
+  {
+    case EQUAL: return rewriteEqual(n);
+    case BAG_IS_INCLUDED: return rewriteIsIncluded(n);
+    default: break;
+  }
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
+RewriteResponse BagsRewriter::rewriteEqual(const TNode& n) const
+{
+  Assert(n.getKind() == EQUAL);
+  if (n[0] == n[1])
+  {
+    // (= A A) = true where A is a bag
+    NodeManager* nm = NodeManager::currentNM();
+    return RewriteResponse(REWRITE_DONE, nm->mkConst(true));
+  }
+  if (n[0].getKind() == MK_BAG && n[1].getKind() == MK_BAG)
```

Comment:
We don't want this rewrite.  Equalities (= t s) should either be rewritten to true, false, (= t s), or (= s t) due to the way theory combination works.  This violates this rule.

Also, we generally never introduce ITE in the rewriter, since this makes things harder to reason about.

Reply:
Done by removing the ITE rule. Probably we would need this [ite commit](https://github.com/mudathirmahgoub/CVC4/commit/4c74f3357710cc11994446f735a59387777d52b4) later in the solver.